hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/move/MoveManagerImpl.java

```diff
@@ -0,0 +1,777 @@
+        //TimeoutException here is caused by ratis request timeout.
+        //we do not throw TimeoutException out here since completeMove is called
+        //in many places. Handling this exception inside itself seems
+        //a better way to have lower complexity.
+        LOG.warn("fail to delete move from DB");
+      }
+
+      // no matter deletion is successful or not, we could complete the
+      // corresponding pending CompletableFuture, because :
+      // 1 if completeMove is called , it means move is actually completed.
+      // 2 a CompletableFuture here means a `move` method call from
+      // container balancer, and balancer will block at CompletableFuture.get(),
+      // so we would better let it return as soon as possible.
+      CompletableFuture<MoveResult> future = pendingMoveFuture.get(cid);
+      pendingMoveFuture.remove(cid);
+      if (successful) {
+        pendingMoveOps.remove(cid);
+        if (future != null && mr != null) {
+          // when we know the future is null, and we want to complete
+          // the move , then we set mr to null.
+          // for example , after becoming a new leader, we might need
+          // to complete some moves and we know these futures does not
+          // exist.
+          future.complete(mr);
+        }
+      } else {
+        //TODO: if unsuccessful(actually rarely), cid will not be removed
+        // from pendingMoveOps. we need to try to delete cid from db and
+        // remove it from pendingMoveOps again. we could start a background
+        // service to handle those failed cases.
+        // Note that, when scm leader switches, new leader will also try
+        // to delete those stale completed move from db.
+        if (future != null) {
+          future.complete(MoveResult.FAIL_CAN_NOT_RECORD_TO_DB);
+        }
+      }
+    }
+  }
+
+  /**
+   * start a move for a given container.
+   *
+   * @param cid Container id
+   * @param src source datanode
+   * @param tgt target datanode
+   */
+  private void startMove(
+      final ContainerID cid, final DatanodeDetails src,
+      final DatanodeDetails tgt, final CompletableFuture<MoveResult> ret)
+      throws TimeoutException, ContainerReplicaNotFoundException,
+      ContainerNotFoundException {
+    if (!pendingMoveOps.containsKey(cid)) {
+      MoveDataNodePair mp = new MoveDataNodePair(src, tgt);
+      if (moveDBHandler.addToDB(cid, mp)) {
+        sendReplicateCommand(cid, tgt, src);
+        pendingMoveOps.putIfAbsent(cid, mp);
+        pendingMoveFuture.putIfAbsent(cid, ret);
+      } else {
+        ret.complete(MoveResult.FAIL_CAN_NOT_RECORD_TO_DB);
+      }
+    }
+  }
+
+  /**
+   * notify MoveManager that the current scm has become leader and ready.
+   */
+  @Override
+  public void onLeaderReady() {
+    //discard all stale records
+    pendingMoveOps.clear();
+    pendingMoveFuture.clear();
+
+    //initialize with db record
+    try (TableIterator<ContainerID,
```

Comment:
When leader is ready, check if records are flushed as present in transaction buffer on switch to leader. Otherwise table will not have any records.

Reply:
yes， this is really a problem, thanks for pointing out, will fix it